File: sdks/go/cmd/elydora/plugins/cline_installation.go
package plugins

import "fmt"

type clineRuntimePaths struct {
	managedRuntimePaths
	hooks clineHookPaths
}

func clineAgentPaths(config InstallConfig, hooks clineHookPaths) (*clineRuntimePaths, error) {
	if err := validateManagedInstallConfig(config, clineAgentKey, "Cline"); err != nil {
		return nil, err
	}
	paths, err := resolveManagedRuntimePaths(config, clineGuardScript, clineAuditScript)
	if err != nil {
		return nil, err
	}
	return &clineRuntimePaths{managedRuntimePaths: *paths, hooks: hooks}, nil
}

func preflightClineInstallation(
	config InstallConfig,
	hooks clineHookPaths,
) (*clineRuntimePaths, error) {
	paths, err := clineAgentPaths(config, hooks)
	if err != nil {
		return nil, err
	}
	if err := validateManagedRuntimeIdentity(
		paths.agentDirectory, config.AgentID, clineAgentKey, "Cline",
	); err != nil {
		return nil, err
	}
	if _, err := resolveAbsoluteNodeRuntime("Cline"); err != nil {
		return nil, err
	}
	return paths, nil
}

func clineExpectedScripts(agentID string, _ map[string]any) ([]byte, []byte) {
	return []byte(generateGuardScript(clineAgentKey, agentID, "", false, "")),
		[]byte(buildHookScriptWithOutput(clineAgentKey, agentID, "", false, true))
}

func prepareClineInstallationChanges(
	config InstallConfig,
	paths *clineRuntimePaths,
	guardFile clineHookFile,
	auditFile clineHookFile,
) ([]*fileChange, error) {
	if paths == nil {
		return nil, fmt.Errorf("prepared Cline installation is required")
	}
	if err := validateManagedRuntimeIdentity(
		paths.agentDirectory, config.AgentID, clineAgentKey, "Cline",
	); err != nil {
		return nil, err
	}
	guardMetadata, err := buildClineMetadata("guard", config.AgentID, paths.guardPath)
	if err != nil {
		return nil, fmt.Errorf("build Cline guard metadata: %w", err)
	}
	auditMetadata, err := buildClineMetadata("audit", config.AgentID, paths.auditPath)
	if err != nil {
		return nil, fmt.Errorf("build Cline audit metadata: %w", err)
	}
	guardWrapper, err := buildClineWrapper(guardMetadata)
	if err != nil {
		return nil, fmt.Errorf("build Cline guard wrapper: %w", err)
	}
	auditWrapper, err := buildClineWrapper(auditMetadata)
	if err != nil {
		return nil, fmt.Errorf("build Cline audit wrapper: %w", err)
	}
	if _, err := clineContractForFiles(
		clineHookFile{exists: true, filePath: guardFile.filePath, source: guardWrapper, metadata: &guardMetadata},
		clineHookFile{exists: true, filePath: auditFile.filePath, source: auditWrapper, metadata: &auditMetadata},
	); err != nil {
		return nil, err
	}
	guardScript, auditScript := clineExpectedScripts(config.AgentID, nil)
	changes, err := managedRuntimeFileChanges(
		config, &paths.managedRuntimePaths, clineAgentKey, string(guardScript), string(auditScript),
	)
	if err != nil {
		return nil, err
	}
	for _, item := range []struct {
		file   clineHookFile
		label  string
		source string
	}{
		{guardFile, "Cline PreToolUse hook", guardWrapper},
		{auditFile, "Cline PostToolUse hook", auditWrapper},
	} {
		change, err := prepareSourceChange(
			item.file.filePath,
			item.label,
			[]byte(item.file.source),
			item.file.exists,
			[]byte(item.source),
			0700,
			false,
		)
		if err != nil {
			return nil, err
		}
		changes = append(changes, change)
	}
	return changes, nil
}

File: sdks/node/src/plugins/cline-installation.ts
import path from 'node:path';
import type { InstallConfig } from './base.js';
import {
  AGENT_KEY,
  AUDIT_SCRIPT,
  GUARD_SCRIPT,
  assertWrapperIntegrity,
  buildMetadata,
  buildWrapper,
  type ClineHookFile,
  runtimeContract,
  sameAgentId,
} from './cline-contract.js';
import {
  commitManagedInstallation,
  preflightManagedInstallation,
  prepareManagedInstallation,
  type ManagedRuntimePaths,
  type PreparedManagedInstallation,
  type RenameFile,
} from './managed-installation.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
  type PreparedManagedTransaction,
} from './managed-transaction.js';

const DISPLAY_NAME = 'Cline';
const DIRECTORY_LABEL = 'Cline hooks directory';

export type ClineRuntimePaths = ManagedRuntimePaths;
export type PreparedClineInstallation = PreparedManagedInstallation;
export type { RenameFile };

function hookLocations(files: readonly ClineHookFile[]) {
  return files.map((file) => ({
    directoryLabel: DIRECTORY_LABEL,
    filePath: file.filePath,
  }));
}

export async function preflightClineInstallation(
  config: InstallConfig,
  files: readonly ClineHookFile[],
): Promise<ClineRuntimePaths> {
  return preflightManagedInstallation({
    agentKey: AGENT_KEY,
    hookLocations: hookLocations(files),
    config,
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
}

export async function prepareClineInstallation(
  config: InstallConfig,
  guardFile: ClineHookFile,
  auditFile: ClineHookFile,
): Promise<PreparedClineInstallation> {
  const paths = await preflightClineInstallation(config, [guardFile, auditFile]);
  const guardMetadata = buildMetadata('guard', config.agentId, paths.guardPath);
  const auditMetadata = buildMetadata('audit', config.agentId, paths.auditPath);
  const guardSource = buildWrapper(guardMetadata);
  const auditSource = buildWrapper(auditMetadata);
  runtimeContract(
    { exists: true, filePath: guardFile.filePath, source: guardSource, metadata: guardMetadata },
    { exists: true, filePath: auditFile.filePath, source: auditSource, metadata: auditMetadata },
  );
  return prepareManagedInstallation({
    agentKey: AGENT_KEY,
    displayName: DISPLAY_NAME,
    hookSources: [{
      directoryLabel: DIRECTORY_LABEL,
      label: 'Cline PreToolUse hook',
      filePath: guardFile.filePath,
      expectedSource: guardFile.source,
      source: guardSource,
    }, {
      directoryLabel: DIRECTORY_LABEL,
      label: 'Cline PostToolUse hook',
      filePath: auditFile.filePath,
      expectedSource: auditFile.source,
      source: auditSource,
    }],
    config,
    auditOptions: { nativePayload: true },
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
}

export async function commitClineInstallation(
  prepared: PreparedClineInstallation,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedInstallation(prepared, renameFile);
}

export async function prepareClineUninstall(
  files: readonly ClineHookFile[],
  agentId?: string,
): Promise<PreparedManagedTransaction> {
  const owned = files.filter((file) => {
    const ownedAgentId = file.metadata?.agentId;
    return ownedAgentId !== undefined
      && (agentId === undefined || sameAgentId(ownedAgentId, agentId));
  });
  for (const file of owned) assertWrapperIntegrity(file);
  const changes = await Promise.all(owned.map((file) => prepareManagedFileChange({
    filePath: file.filePath,
    label: `Cline ${file.metadata!.kind} hook`,
    next: undefined,
    mode: 0o600,
    expectedSource: file.source,
    verifyExpectedSource: true,
  })));
  return {
    displayName: DISPLAY_NAME,
    operation: 'uninstall',
    directories: owned.map((file) => ({
      path: path.dirname(file.filePath),
      label: DIRECTORY_LABEL,
    })),
    changes: changes.filter((change) => change !== undefined),
  };
}

export async function commitClineUninstall(
  prepared: PreparedManagedTransaction,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedTransaction(prepared, renameFile);
}
